### ISSUE TITLE
Mpirun fails with local peers error

### ISSUE THREAD

**Issue by dipti-kothari** [openpmix/prrte#997]
### What version of the PMIx Reference Server are you using? (e.g., v1.0, v2.1, git master @ hash, etc.)
Prrte: master @ 9237456

##### What version of PMIx are you using? (e.g., v1.2.5, v2.0.3, v2.1.0, git branch name and hash, etc.)
Openpmix: master @ 10b03e44

### Please describe the system on which you are running

* Operating system/version: alinux2
* Computer hardware: AWS c5n.18xlarge instance
* Network type: EFA

-----------------------------

## Details of the problem

Running osu latency over two nodes, mpirun fails with unable to retrieve local peers. Hostfile consists of IP addresses and this is not a managed environment. The issue started happening after prrte commit d2c6ef6313

```
[ec2-user@ip-172-31-62-180 ~]$ mpirun --prefix /home/ec2-user/ompi/install/ -n 2 --map-by ppr:1:node --output tag  --hostfile hostfile /home/ec2-user/osu-micro-benchmarks-5.6.3/libexec/osu-micro-benchmarks/mpi/pt2pt/osu_latency
[1,1]<stderr>:--------------------------------------------------------------------------
[1,1]<stderr>:It looks like MPI runtime init failed for some reason; your parallel process is
[1,1]<stderr>:likely to abort.  There are many reasons that a parallel process can
[1,1]<stderr>:fail during RTE init; some of which are due to configuration or environment
[1,1]<stderr>:problems.  This failure appears to be an internal failure; here's some
[1,1]<stderr>:additional information (which may only be relevant to an Open MPI
[1,1]<stderr>:developer):
[1,1]<stderr>:
[1,1]<stderr>:  local peers
[1,1]<stderr>:  --> Returned "Not found" (-13) instead of "Success" (0)
[1,1]<stderr>:--------------------------------------------------------------------------
[1,1]<stderr>:--------------------------------------------------------------------------
[1,1]<stderr>:It looks like MPI_INIT failed for some reason; your parallel process is
[1,1]<stderr>:likely to abort.  There are many reasons that a parallel process can
[1,1]<stderr>:fail during MPI_INIT; some of which are due to configuration or environment
[1,1]<stderr>:problems.  This failure appears to be an internal failure; here's some
[1,1]<stderr>:additional information (which may only be relevant to an Open MPI
[1,1]<stderr>:developer):
[1,1]<stderr>:
[1,1]<stderr>:  ompi_mpi_init: ompi_rte_init failed
[1,1]<stderr>:  --> Returned "Not found" (-13) instead of "Success" (0)
[1,1]<stderr>:--------------------------------------------------------------------------
[1,1]<stderr>:*** An error occurred in MPI_Init
[1,1]<stderr>:*** on a NULL communicator
[1,1]<stderr>:*** MPI_ERRORS_ARE_FATAL (processes in this communicator will now abort,
[1,1]<stderr>:***    and MPI will try to terminate your MPI job as well)
[1,1]<stderr>:[172.31.56.61:02508] Local abort before MPI_INIT completed completed successfully, but am not able to aggregate error messages, and not able to guarantee that all other processes were killed!
--------------------------------------------------------------------------
mpirun detected that one or more processes exited with non-zero status, thus causing
the job to be terminated. The first process to do so was:

  Process name: [mpirun-ip-172-31-62-180-1373@1,1]
  Exit code:    14
--------------------------------------------------------------------------
```

**Comment by dancejic** [comment]
We saw this in our nightly tests, I'm interested in this line in particular:
`[compute-st-c5n18xlarge-56:32026] [[57509,1],3384] selected pml ob1, but peer [[57509,1],0] on compute-st-c5n18xlarge-116 selected pml cm`

this is the full error message:

```
[compute-st-c5n18xlarge-56:32026] [[57509,1],3384] selected pml ob1, but peer [[57509,1],0] on compute-st-c5n18xlarge-116 selected pml cm
--------------------------------------------------------------------------
MPI_INIT has failed because at least one MPI process is unreachable
from another.  This *usually* means that an underlying communication
plugin -- such as a BTL or an MTL -- has either not loaded or not
allowed itself to be used.  Your MPI job will now abort.

You may wish to try to narrow down the problem;

 * Check the output of ompi_info to see which BTL/MTL plugins are
   available.
 * Run your application with MPI_THREAD_SINGLE.
 * Set the MCA parameter btl_base_verbose to 100 (or mtl_base_verbose,
   if using MTL-based communications) to see exactly which
   communication plugins were considered and/or discarded.
--------------------------------------------------------------------------
[compute-st-c5n18xlarge-56:32026] *** An error occurred in MPI_Init_thread
[compute-st-c5n18xlarge-56:32026] *** reported by process [3768909825,3384]
[compute-st-c5n18xlarge-56:32026] *** on a NULL communicator
[compute-st-c5n18xlarge-56:32026] *** Unknown error
[compute-st-c5n18xlarge-56:32026] *** MPI_ERRORS_ARE_FATAL (processes in this communicator will now abort,
[compute-st-c5n18xlarge-56:32026] ***    and potentially your MPI job)
--------------------------------------------------------------------------
The OFI MTL was not able to find endpoint information for a remote
endpoint.  Most likely, this means that the remote process was unable
to initialize the Libfabric NIC correctly.  This error is not
recoverable and your application is likely to abort.

  Local host: compute-st-c5n18xlarge-56
  Remote host: compute-st-c5n18xlarge-56
  Error: Not found (-13)
--------------------------------------------------------------------------
```

**Comment by rhc54** [comment]
I believe @awlauria has a PR coming that will probably address the "local peers" issue. The `pml` mismatch has nothing to do with PRRTE - that's purely an OMPI problem and should be raised there.

**Comment by dancejic** [comment]
That was my mistake, that's a separate issue.

**Comment by rhc54** [comment]
Ah...I know why you didn't get any local peers - you don't have any!! The local peers contains the ranks of all peers from that job that share a node with you - _excluding_ yourself. So if you are all alone on a node, then you don't have any local peers.

So this is a problem in OMPI, not PRRTE. We need to update the OMPI code to correctly handle the situation of being the only proc from a job on a node. I'll take care of that and have it close this PR.

**Comment by dipti-kothari** [comment]
I increased the number of processes per node and I am still getting the same error
```
[ec2-user@ip-172-31-62-180 ~]$ mpirun --prefix /home/ec2-user/ompi/install/ -n 72 --map-by ppr:36:node --output tag  --hostfile /home/ec2-user/hostfile /home/ec2-user/osu-micro-benchmarks-5.6.3/libexec/osu-micro-benchmarks/mpi/collective/osu_allgather -x 10 -i 10
[1,42]<stderr>:--------------------------------------------------------------------------
[1,42]<stderr>:It looks like MPI runtime init failed for some reason; your parallel process is
[1,42]<stderr>:likely to abort.  There are many reasons that a parallel process can
[1,42]<stderr>:fail during RTE init; some of which are due to configuration or environment
[1,42]<stderr>:problems.  This failure appears to be an internal failure; here's some
[1,42]<stderr>:additional information (which may only be relevant to an Open MPI
[1,42]<stderr>:developer):
[1,42]<stderr>:
[1,42]<stderr>:  local peers
[1,42]<stderr>:  --> Returned "Not found" (-13) instead of "Success" (0)
[1,42]<stderr>:--------------------------------------------------------------------------
[1,40]<stderr>:--------------------------------------------------------------------------
[1,40]<stderr>:It looks like MPI runtime init failed for some reason; your parallel process is
[1,40]<stderr>:likely to abort.  There are many reasons that a parallel process can
[1,40]<stderr>:fail during RTE init; some of which are due to configuration or environment
[1,40]<stderr>:problems.  This failure appears to be an internal failure; here's some
[1,40]<stderr>:additional information (which may only be relevant to an Open MPI
[1,40]<stderr>:developer):
[1,40]<stderr>:
[1,40]<stderr>:  local peers
[1,40]<stderr>:  --> Returned "Not found" (-13) instead of "Success" (0)
[1,40]<stderr>:--------------------------------------------------------------------------
[1,36]<stderr>:--------------------------------------------------------------------------
....<trunc>
[1,71]<stderr>:--------------------------------------------------------------------------
[1,71]<stderr>:*** An error occurred in MPI_Init
[1,71]<stderr>:*** on a NULL communicator
[1,71]<stderr>:*** MPI_ERRORS_ARE_FATAL (processes in this communicator will now abort,
[1,71]<stderr>:***    and MPI will try to terminate your MPI job as well)
[1,71]<stderr>:[172.31.56.61:01724] Local abort before MPI_INIT completed completed successfully, but am not able to aggregate error messages, and not able to guarantee that all other processes were killed!
[ip-172-31-62-180.us-west-2.compute.internal:14099] pmix_ptl_base: send_msg: write failed: Connection reset by peer (104) [sd = 92]
[ip-172-31-62-180.us-west-2.compute.internal:14099] pmix_ptl_base: send_msg: write failed: Connection reset by peer (104) [sd = 102]
[ip-172-31-62-180.us-west-2.compute.internal:14099] pmix_ptl_base: send_msg: write failed: Connection reset by peer (104) [sd = 140]
[ip-172-31-62-180.us-west-2.compute.internal:14099] pmix_ptl_base: send_msg: write failed: Connection reset by peer (104) [sd = 119]
--------------------------------------------------------------------------
mpirun detected that one or more processes exited with non-zero status, thus causing
the job to be terminated. The first process to do so was:

  Process name: [mpirun-ip-172-31-62-180-14099@1,36]
  Exit code:    14
--------------------------------------------------------------------------
```

**Comment by rhc54** [comment]
And this is with current OMPI master?

**Comment by dipti-kothari** [comment]
Yes, it is with the current ompi master

**Comment by rhc54** [comment]
This is odd - it is working fine for me on my 3-box Linux cluster with the current head of OMPI master. I am seeing a different issue on my Mac that I'll have to chase down. But I cannot replicate this reported problem.

What if you try running a simple "hello" such as the examples/hello_c at the top of the OMPI tree? Just trying to cut down some of the possible complications - want to ensure that the application was compiled against this version of OMPI.

**Comment by dipti-kothari** [comment]
I tried it with hello_c, still the same error
```
[ec2-user@ip-172-31-62-180 ~]$ mpirun --prefix /home/ec2-user/ompi/install/ -n 2 --map-by ppr:1:node --output tag  --hostfile /home/ec2-user/hostfile /home/ec2-user/ompi/examples/hello_c
[1,1]<stderr>:--------------------------------------------------------------------------
[1,1]<stderr>:It looks like MPI runtime init failed for some reason; your parallel process is
[1,1]<stderr>:likely to abort.  There are many reasons that a parallel process can
[1,1]<stderr>:fail during RTE init; some of which are due to configuration or environment
[1,1]<stderr>:problems.  This failure appears to be an internal failure; here's some
[1,1]<stderr>:additional information (which may only be relevant to an Open MPI
[1,1]<stderr>:developer):
[1,1]<stderr>:
[1,1]<stderr>:  local peers
[1,1]<stderr>:  --> Returned "Not found" (-13) instead of "Success" (0)
[1,1]<stderr>:--------------------------------------------------------------------------
[1,1]<stderr>:--------------------------------------------------------------------------
[1,1]<stderr>:It looks like MPI_INIT failed for some reason; your parallel process is
[1,1]<stderr>:likely to abort.  There are many reasons that a parallel process can
[1,1]<stderr>:fail during MPI_INIT; some of which are due to configuration or environment
[1,1]<stderr>:problems.  This failure appears to be an internal failure; here's some
[1,1]<stderr>:additional information (which may only be relevant to an Open MPI
[1,1]<stderr>:developer):
[1,1]<stderr>:
[1,1]<stderr>:  ompi_mpi_init: ompi_rte_init failed
[1,1]<stderr>:  --> Returned "Not found" (-13) instead of "Success" (0)
[1,1]<stderr>:--------------------------------------------------------------------------
[1,1]<stderr>:*** An error occurred in MPI_Init
[1,1]<stderr>:*** on a NULL communicator
[1,1]<stderr>:*** MPI_ERRORS_ARE_FATAL (processes in this communicator will now abort,
[1,1]<stderr>:***    and MPI will try to terminate your MPI job as well)
[1,1]<stderr>:[172.31.56.61:14331] Local abort before MPI_INIT completed completed successfully, but am not able to aggregate error messages, and not able to guarantee that all other processes were killed!
--------------------------------------------------------------------------
```
I can print the hostname, though
```
[ec2-user@ip-172-31-62-180 ~]$ mpirun --prefix /home/ec2-user/ompi/install/ -n 2 --map-by ppr:1:node --output tag  --hostfile /home/ec2-user/hostfile hostname
[1,0]<stdout>:ip-172-31-62-180.us-west-2.compute.internal
[1,1]<stdout>:ip-172-31-56-61.us-west-2.compute.internal
```

**Comment by rhc54** [comment]
What is in your hostfile?

No surprise on running hostname as it isn't looking for any PMIx information. I'm searching for the difference in our systems. Could be in the way the hosts are expressed.

**Comment by rhc54** [comment]
Sigh...so this is another edge case of the hostfile not matching the host names. I'm always amazed at how many ways the computing community has come up with to create problems for themselves 😦 

The problem is that we aren't recognizing the hosts in the hostfile as matching any of the hosts we are launching on. So as far as PRRTE is concerned, you have no local peers because there is nothing on this node. We'll have to figure out how to get the nodes to be recognized _without_ doing a DNS resolve since the DNS is too bloody slow.

What a mess...

**Comment by rhc54** [comment]
@dipti-kothari Could you please try again with the latest head of PMIx and PRRTE master?

**Comment by rhc54** [comment]
Was finally able to replicate on my machine - fix is in #1005
